internal/provider/serverless_cluster_resource.go

```diff
@@ -0,0 +1,730 @@
+			},
+		}
+	}
+
+	data.Endpoints = &endpoints{
+		PublicEndpoint: publicEndpoint{
+			Host:     types.StringValue(*e.Public.Host),
+			Port:     types.Int64Value(int64(*e.Public.Port)),
+			Disabled: types.BoolValue(*e.Public.Disabled),
+		},
+		PrivateEndpoint: pe,
+	}
+
+	en := resp.EncryptionConfig
+	data.EncryptionConfig = &encryptionConfig{
+		EnhancedEncryptionEnabled: types.BoolValue(*en.EnhancedEncryptionEnabled),
+	}
+
+	data.HighAvailabilityType = types.StringValue(string(*resp.HighAvailabilityType))
+	data.Version = types.StringValue(*resp.Version)
+	data.CreatedBy = types.StringValue(*resp.CreatedBy)
+	data.CreateTime = types.StringValue(resp.CreateTime.String())
+	data.UpdateTime = types.StringValue(resp.UpdateTime.String())
+	data.UserPrefix = types.StringValue(*resp.UserPrefix)
+	data.State = types.StringValue(string(*resp.State))
+
+	u := resp.Usage
+	data.Usage = &usage{
+		RequestUnit:     types.StringValue(*u.RequestUnit),
+		RowBasedStorage: types.Int64Value(int64(*u.RowBasedStorage)),
+		ColumnarStorage: types.Int64Value(int64(*u.ColumnarStorage)),
+	}
+
+	data.Labels = labels
+	data.Annotations = annotations
+}
+
+func WaitServerlessClusterReady(ctx context.Context, timeout time.Duration, interval time.Duration, clusterId string,
+	client tidbcloud.TiDBCloudServerlessClient) (*clusterV1beta1.TidbCloudOpenApiserverlessv1beta1Cluster, error) {
+	stateConf := &retry.StateChangeConf{
+		Pending: []string{
+			string(serverlessClusterStatusCreating),
+			string(serverlessClusterStatusDeleting),
+			string(serverlessClusterStatusRestoring),
+			string(serverlessClusterStatusInactive),
+			string(serverlessClusterStatusUpgrading),
+			string(serverlessClusterStatusImporting),
+			string(serverlessClusterStatusModifying),
+			string(serverlessClusterStatusPausing),
+			string(serverlessClusterStatusResuming),
+		},
+		Target: []string{
+			string(serverlessClusterStatusActive),
+			string(serverlessClusterStatusPaused),
+			string(serverlessClusterStatusDeleted),
+			string(serverlessClusterStatusMaintenance),
+		},
+		Timeout:      timeout,
+		MinTimeout:   500 * time.Millisecond,
+		PollInterval: interval,
+		Refresh:      serverlessClusterStateRefreshFunc(ctx, clusterId, client),
+	}
+
+	outputRaw, err := stateConf.WaitForStateContext(ctx)
+
+	if output, ok := outputRaw.(*clusterV1beta1.TidbCloudOpenApiserverlessv1beta1Cluster); ok {
+		return output, err
+	}
+	return nil, err
+}
+
+func serverlessClusterStateRefreshFunc(ctx context.Context, clusterId string,
+	client tidbcloud.TiDBCloudServerlessClient) retry.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		tflog.Trace(ctx, "Waiting for serverless cluster ready")
+		cluster, err := client.GetCluster(ctx, clusterId, clusterV1beta1.SERVERLESSSERVICEGETCLUSTERVIEWPARAMETER_FULL)
+		if err != nil {
+			return nil, "", err
+		}
+		return cluster, string(*cluster.State), nil
+	}
+}
```

Comment:
Add import for serverless cluster resources, it has been useful during the past experience